src/chunk.c

```diff
@@ -1051,6 +1186,183 @@ chunk_find_all_oids(Hyperspace *hs, List *dimension_vecs, LOCKMODE lockmode)
+				 errmsg("older_than_type and newer_than_type should have the same type")
+				 ));
+
+	/*
+	 * Cache outside the if block to make sure cached hypertable entry
+	 * returned will still be valid in foreach block below
+	 */
+	hypertable_cache = hypertable_cache_pin();
+	if (PG_ARGISNULL(0))
+	{
+		hypertables = hypertable_get_all();
+	}
+	else
+	{
+		ht = hypertable_cache_get_entry(hypertable_cache, table_relid);
+		if (!ht)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("table \"%s\" does not exist or is not a hypertable",
+							get_rel_name(table_relid))
+					 ));
+		hypertables = list_make1(ht);
+	}
+
+	oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
+	ctxs = palloc(sizeof(ChunkScanCtx *) * list_length(hypertables));
+	MemoryContextSwitchTo(oldcontext);
+	foreach(lc, hypertables)
+	{
+		ht = lfirst(lc);
+
+		time_dim = hyperspace_get_open_dimension(ht->space, 0);
+
+		if (time_dim_type == InvalidOid)
+			time_dim_type = time_dim->fd.column_type;
+
+		/*
+		 * Even though internally all time columns are represented as bigints,
+		 * it is locally unclear what set of chunks should be returned if
+		 * there are multiple tables on the system some of which care about
+		 * timestamp when others do not. That is why, whenever there is any
+		 * time dimension constraint given as an argument (older_than or
+		 * newer_than) we make sure all hypertables have the time dimension
+		 * type of the given type or through an error. This check is done
+		 * accross hypertables that is why it is not in the helper function
+		 * below.
+		 */
+		if (time_dim_type != time_dim->fd.column_type &&
+			(older_than_type != InvalidOid || newer_than_type != InvalidOid))
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("cannot call \"%s\" on all hypertables "
+							"when all hypertables do not have the same time dimension type",
+							caller_name)
+					 ));
+
+		ctxs[ht_index++] = chunks_typecheck_and_find_all_in_range_limit(ht->space,
+																		time_dim,
+																		older_than_datum,
+																		older_than_type,
+																		newer_than_datum,
+																		newer_than_type,
+																		-1,
+																		funcctx->multi_call_memory_ctx,
+																		caller_name,
+																		&num_chunks);
+
+	}
+	oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
+	/* Q:: two options here: */
+	/* have chunks to be chunks and copy chunk structs over */
+	/* have chunks be Oids or a new lighter datatype */
+	/* First one has more overhead but makes ordering the results much easier */
+	/* as we can use hypertable d and table id. */
+
+	/*
+	 * The first one is more effient and makes sure we do not deal with mem
+	 * context switches that much
+	 */
+
+	/*
+	 * num_chunks can safely be 0 as palloc protects against unportable
+	 * behavior.
+	 */
+	chunks = palloc(sizeof(Chunk *) * num_chunks);
```

Comment:
Should I allocate for Chunk* pointers here? Note that this is called after hash context destroy. It looks like hash context destroy only affects ctx->htab and this allocation works so Chunks are not freed.
But I am not sure where exactly are Chunks structs freed. There is a function for doing that but is never called.
@erimatnor just wanted to draw your attention on this to make sure this is not going to result in some kind of leak.

Reply:
There is the option of actually allocating `Chunk` structs as opposed to pointers and the behavior in that case will definitely be clearer to understand from the code. However, `Chunk`s are  pretty big.
On the other hand, we could make it an array of `Oid`s and it would actually simplify and generalize the `chunks_return_srf ` function. But this becomes problematic if we want to sort the oids in the order defined by `chunk_cmp` as it uses other info from `Chunk` for sorting.